src/nagini_translation/translators/expression.py

```diff
@@ -717,6 +717,46 @@ def translate_static_field_access(self, field: PythonGlobalVar,
                 ctx.current_function.call_deps.add((node, receiver, ctx.module))
         return res
 
+    def translate_adt_decons(self, decons: PythonClass, node: ast.Attribute,
+                             pos: Position, ctx: Context) -> StmtsAndExpr:
+        """
+        Deconstructs the ADT into the selected component, boxing the result
+        if the projected component's type is ADT (recursive composition).
+        """
+        info = self.no_info(ctx)
+        adt_name = decons.adt_def.name + '_ADT'
+        adt_prefix = decons.adt_def.name + '_'
+        adt_type = self.viper.DomainType(adt_name, {}, [])
+
+        # If ADT is casted, ignore the casting since 'decons' already reflects
```

Comment:
Why is this necessary? Or is it an optimization?

Reply:
I added some postconditions to box ADT function so that this is no longer necessary and the regular cast verification is being generated again.

Reply:
Great, thanks. I think this is the more general solution.